graphql-server/src/index.ts

```diff
@@ -39,14 +54,202 @@ async function main() {
+    passport.authenticate("oidc", {
+      successRedirect: "/post-login",
+      failureRedirect: "/",
+    })(req, res, next);
+  });
+
+  const checkAuthenticated = (req: any, res: any, next: any) => {
+    if (loginTime !== 0) {
+      if (Date.now() - loginTime > ssoSessionIdleTimeout) {
+        // Clear Passport session/user object from req
+        req.logOut((error: any) => {
+          if (error) {
+            return next(error);
+          }
+        });
+      } else {
+        loginTime = Date.now();
+      }
+    }
+
+    if (req.isAuthenticated()) {
+      return next();
+    } else {
+      res.status(401).send("401 Unauthorized");
+    }
+  };
+
+  app.get("/post-login", checkAuthenticated, (req: any, res) => {
+    loginTime = Date.now();
+
+    const userEmail = req.user.email;
+    res.send(`
+      <script>
+        window.opener.postMessage(${JSON.stringify(
+          userEmail
+        )}, "https://localhost:3006/patients");
+        window.onload = function() {
+          setTimeout(function() {
+            window.close();
+          }, 1000);
+        };
+      </script>
+      <p>You are logged in.</p>
+    `);
+  });
+
+  app.post("/logout", (req: any, res, next) => {
+    req.logOut((error: any) => {
+      if (error) {
+        return next(error);
+      }
+    });
+
+    // Clear Keycloak session directly without "log out?" prompt
+    res.redirect(
+      keycloakClient.endSessionUrl({
+        id_token_hint: id_token,
+      })
+    );
+  });
+
+  app.get("/check-login", checkAuthenticated, async (req: any, res) => {
+    res.status(200).send(req.user.email);
+  });
+
+  function checkAuthorized(req: any, res: any, next: any) {
+    const userRoles = req.user.groups;
+    if (userRoles.includes("mrn-search")) {
+      return next();
+    } else {
+      res.status(403).send("403 Forbidden");
+    }
+  }
+
+  const logDir = path.join(process.env.SMILE_DATA_HOME, props.log_dir);
+  if (!fs.existsSync(logDir)) fs.mkdirSync(logDir);
+  const accessLogStream = fs.createWriteStream(path.join(logDir, "event.log"), {
+    flags: "a+",
+  });
+  morgan.token("userKeycloakId", function (req: any, res: any) {
+    return req.user ? req.user.sub : "-";
+  });
+  app.use(
+    morgan(":method :url :status - :date[iso] - :userKeycloakId", {
+      stream: accessLogStream,
+      skip: function (req: any, res: any) {
+        return req.path !== "/mrn-search";
+      },
+    })
+  );
+
+  app.post(
+    "/mrn-search",
+    checkAuthenticated,
+    checkAuthorized,
+    async (req, res) => {
+      const patientMrns = req.body;
+      const patientIdsTriplets = [];
+
+      if (os.arch() !== "arm64" && oracledb !== null) {
+        const connection = await oracledb.getConnection({
+          user: props.oracle_user,
+          password: props.oracle_password,
+          connectString: props.oracle_connect_string,
+        });
+
+        for (const patientMrn of patientMrns) {
```

Comment:
this code can be improved using Promise.all.  We can discuss.

Reply:
I got a head start and thought the refactored code could look like the below. Will save this for when we chat live.

```
const patientIdsTriplets = [];

if (os.arch() !== "arm64" && oracledb !== null) {
  const connection = await oracledb.getConnection({
    user: props.oracle_user,
    password: props.oracle_password,
    connectString: props.oracle_connect_string,
  });

  const promises = patientMrns.map(async (patientMrn) => {
    const result = await connection.execute(
      "SELECT CMO_ID, DMP_ID, PT_MRN FROM CRDB_CMO_LOJ_DMP_MAP WHERE :patientMrn IN (DMP_ID, PT_MRN, CMO_ID)",
      { patientMrn }
    );
    if (result.rows.length > 0) {
      return result.rows[0];
    }
  });

  patientIdsTriplets.push(...(await Promise.all(promises)));

  await connection.close();
}
```